README.md
# fluent_query_builder

[![Build Status](https://api.travis-ci.org/insinfo/fluent_query_builder.svg?branch=master)](https://travis-ci.com/github/insinfo/fluent_query_builder) 

[![CI](https://github.com/insinfo/fluent_query_builder/actions/workflows/dart.yml/badge.svg)](https://github.com/insinfo/fluent_query_builder/actions/workflows/dart.yml)

A dart library that allows you to execute SQL queries in the PostgreSQL and MySql database in a fluent way, is very easy to execute.
This library implements POOL of connections.
This lib implements the vast majority of SQL statements and clauses

Soon it will also support ORM without reflection and without code generation in a simple and consistent way.

## Usage

A simple query builder usage example:

```dart
import 'package:fluent_query_builder/fluent_query_builder.dart';

void main() {
  //PostgreSQL connection information
  final pgsqlCom = DBConnectionInfo(
    host: 'localhost',
    database: 'banco_teste',
    driver: ConnectionDriver.pgsql,
    port: 5432,
    username: 'user',
    password: 'pass',
    charset: 'utf8',
    schemes: ['public'],
  );

  //MySQL connection information
  final mysqlCom = DBConnectionInfo(
    host: 'localhost',
    database: 'banco_teste',
    driver: ConnectionDriver.mysql,
    port: 3306,
     username: 'user',
    password: 'pass',
    charset: 'utf8',
  );

  DbLayer().connect(mysqlCom).then((db) {
    //mysql insert
    db
        .insert()
        .into('pessoas')
        .set('nome', 'Isaque Neves Sant\'Ana')
        .set('telefone', '(22) 2771-6265')
        .exec()
        .then((result) => print('mysql insert $result'));

    //mysql update
    db
        .update()
        .table('pessoas')
        .set('nome', 'João')
        .where('id=?', 13)
        .exec()
        .then((result) => print('mysql update $result'));

    //mysql update with setAll
    db
        .update()
        .whereSafe('id', '=', 13)
        .table('pessoas')
        .setAll({
          'nome': 'Jon Doe',
          'telefone': '171171171',
        })
        .exec()
        .then((result) => print('mysql update with setAll $result'));

    //mysql delete
    db.delete().from('pessoas')
    .where('id=?', 14)
    .exec()
    .then((result) => print('mysql delete $result'));

    //mysql select
    db
        .select()
        //.fields(['login', 'idSistema', 's.sigla'])
        //.fieldRaw('SELECT COUNT(*)')
        .from('pessoas')
        .whereSafe('nome', 'like', '%Sant\'Ana%')
        //.limit(1)
        .firstAsMap()
        .then((result) => print('mysql select $result'));

    //mysql raw query SELECT * FROM `pessoas` or SELECT COUNT(*) FROM pessoas
    db  
        .raw("SELECT * FROM `pessoas`")
        .firstAsMap()
        .then((result) => print('mysql raw $result'));
        
    //mysql count records 
    db
        .select()
        .from('pessoas')
        .orWhereSafe('nome', 'like', '%Sant\'Ana%')
        .orWhereSafe('id', '<', 20)
        .count()
        .then((result) => print('mysql count $result'));
  });

  DbLayer().connect(pgsqlCom).then((db) {
    //pgsql insert
    db
        .insert()
        .into('usuarios')
        .set('username', 'isaque')
        .set('password', '123456')
        .exec()
        .then((result) => print('pgsql insert $result'));

  //pgsql insertGetAll
  db
      .insertGetAll()
      .into('usuarios')
      .set('username', 'isaque')
      .set('password', '123456')
      .exec()
      .then((result) => print('pgsql insertGetAll $result'));

 //pgsql insertGetId
  db
      .insertGetId()
      .into('usuarios')
      .set('username', 'isaque')
      .set('password', '123456')
      .exec()
      .then((result) => print('pgsql insertGetId $result'));

    //pgsql count records
         db
        .select()
        .from('pessoas')       
        .count()
        .then((result) => print('pgsql count $result'));

     //Complex selection With whereGroup, whereSafe, where whereRaw
      await db.raw('DROP TABLE IF EXISTS notificacoes').exec();
      await db.raw('''CREATE TABLE notificacoes (
                      "id" serial NOT NULL ,
                      "mensagem" text COLLATE "pg_catalog"."default",
                      "dataCriado" timestamp(0) DEFAULT now(),
                      "link" text COLLATE "pg_catalog"."default" DEFAULT NULL::character varying,
                      "idPessoa" int4,
                      "idSistema" int4,
                      "userAgent" text COLLATE "pg_catalog"."default" DEFAULT NULL::character varying,
                      "idOrganograma" int4,
                      "isLido" bool,
                      "toAll" bool,
                      "icon" text COLLATE "pg_catalog"."default",
                      CONSTRAINT "notificacoes_pkey" PRIMARY KEY ("id")
                    )''').exec();
      await db.insert().into('notificacoes').setAll({
        'mensagem': 'Teste',
        'link': 'https://pub.dev',
        'dataCriado': '2021-05-04 17:53:55',
        'idPessoa': 2,
        'idSistema': 1,
        'userAgent': 'Google',
        'idOrganograma': 19,
        'isLido': false,
        'toAll': true,
        'icon': ''
      }).exec();

      final query = db.select().fromRaw('notificacoes');
      query.where('"dataCriado"::TIMESTAMP  > \'?\'::TIMESTAMP ', '2021-05-04 17:53:55');
      query.whereGroup((q) {
        q.where('"idPessoa"=?', 2, 'or');
        q.where('"idOrganograma"=?', 19, 'or');
        q.where('"toAll"=?', "'t'", 'or');
        q.whereSafe('"toAll"', '=', 'true');
        q.whereRaw('"toAll"= true');
        return q;
      });
      query.orWhereGroup((q) {
        q.whereSafe('"toAll"', '=', 'true');
        query.orWhereGroup((q) {
          q.whereSafe('"toAll"', '=', 'true');
          return q;
        });
        return q;
      });
      query.order('dataCriado', dir: SortOrder.DESC);
      final listMap = await query.limit(1).getAsMap();

      print(listMap)
       /*Resullt
       [
        {
          'id': 1,
          'mensagem': 'Teste',
          'dataCriado': DateTime.tryParse('2021-05-04 17:53:55.000Z'),
          'link': 'https://pub.dev',
          'idPessoa': 2,
          'idSistema': 1,
          'userAgent': 'Google',
          'idOrganograma': 19,
          'isLido': false,
          'toAll': true,
          'icon': ''
        }
      ]*/
   

        
  });
  
  //pgsql transaction example
  var db = await DbLayer().connect(pgsqlCom);
  await db.transaction((ctx) async {
    
    var result = await ctx.insert().into('usuarios')
    .set('username', 'isaque')
    .set('password', '123456')
    .exec();
    //pgsql use of Where Group => WHERE (nome ilike '%5%' or cpf ilike '%5%') AND id > 0
    await ctx
    .select()
    .from('pessoas')     
    .orWhereGroup((query) {
      return query
      .orWhereSafe('nome', 'ilike', '%5%')
      .orWhereSafe('cpf', 'ilike', '%5%');
    })
    .whereSafe('id', '>', 0)
    .getAsMap();

    print('pgsql transaction $result');
  });

  //pgsql getRelationFromMaps example
   var data = await db
      .select()
      .from('pessoas') 
      .orWhereGroup((query) {
        return query.orWhereSafe('nome', 'ilike', '%5%').orWhereSafe('cpf', 'ilike', '%5%');
      })
      .whereSafe('id', '>', 0)
      .getAsMap();

  data = await db.getRelationFromMaps(data, 'usuarios', 'idPessoa', 'id');

  print('pgsql select \r\n ${jsonEncode(data)}');
  /* pgsql select 
    [
      {
        "id":2,
        "nome":"Isaque Neves Sant'Ana",
        "telefone":"(22) 99701-5305",
        "cpf":"54654",
        "usuarios":[
            {
              "id":16,
              "username":"isaque",
              "password":"123456",
              "ativo":null,
              "idPessoa":2
            }
        ]
      },
      {
        "id":1,
        "nome":"João da Silva 5",
        "telefone":"27772339",
        "cpf":"1111",
        "usuarios":[

        ]
      }
    ]  */
}

```

A simple beta pre ORM usage example:

```dart
import 'package:fluent_query_builder/fluent_query_builder.dart';

class Usuario implements FluentModelBase {
  Usuario({this.username});

  Usuario.fromMap(Map<String, dynamic> map) {
    id = map['id'] as int;
    username = map['username'] as String;
    password = map['password'] as String;
    ativo = map['ativo'] as bool;
    idPerfil = map['idPerfil'] as int;
  }

  int id;
  String username;
  String password;
  bool ativo;
  int idPerfil;

  @override
  Map<String, dynamic> toMap() {
    final Map<String, dynamic> data = Map<String, dynamic>();
    if (id != null) {
      data['id'] = id;
    }
    data['username'] = username;
    data['password'] = password;
    data['ativo'] = ativo;
    data['idPerfil'] = idPerfil;
    return data;
  }

  @override
  String get tableName => 'usuarios';
}

void main() {
  //connection settings
  final com = DBConnectionInfo(
    host: '192.168.133.13',
    database: 'sistemas',
    port: 5432,
    username: 'user',
    password: 'pass',
    charset: 'utf8',
    schemes: ['riodasostrasapp'],
  );
  
  //connect on database and set factories for construct instance of model
   DbLayer(factories: [
    {Usuario: (x) => Usuario.fromMap(x)}
  ]).connect(com).then((db) {
    //insert Usuario
    db.putSingle<Usuario>(Usuario(username: 'jon.doe', password: '123456'));
    //update Usuario
    db.update().where('id=?', 20).updateSingle<Usuario>(Usuario(username: 'jon.doe', password: '987'));
    //select Usuario
    db.select().from(Usuario().tableName).where('id>?', 2).fetchAll<Usuario>().then((result) {
      print(result);
    });
    //delete Usuario
    db.delete().deleteSingle<Usuario>(Usuario(id: 20, username: 'jon.doe', password: '123456'));
  });
}

```


CHANGELOG.md
## 1.0.0

- Initial version

## 1.0.1

- fix 

## 1.0.1

- fix dartfmt 

## 1.0.5

- remove static DBLayer db conection

## 1.0.11

- refactor and overall bug fixes

## 1.0.16

- implemented mysql initial support

## 1.1.0

- improvements against SQL injection with the introduction of the whereSafe() and orWhereSafe() methods, introducing the count() method to modify the query to count the database records, in addition to automatically introducing the limit and offset for the first() and firstAsMap(), in addition to several bug fixes for better compatibility with MySQL. And update the Readme with more examples.

## 1.1.1

- fix break in getAsMap method in PostgreSQL implementation

## 1.1.2

- add orWhereGroup and whereGroup method for grouping conditions

## 1.1.8

- implemented insertGetId and insertGetAll methods for returning id and row post inserted

## 1.1.9

- implemented setAll method for insert and update

## 1.1.10

- fix bug on conbine of orWhereGroup and where

## 1.1.11

- fix bug on orderBy

## 1.1.12

- a reconnection method was implemented in MySqlExecutor, so that the application is more resilient in case MySQL or MariaDB or PostgreSQL is restarted

## 1.1.14

- fix bug on where safe, failure on coted col name

## 1.1.16

- fix bug on where 

## 2.0.0

- refactoring for better resiliency in case of loss of connection to the postgresql server, now has an option called "reconnectIfConnectionIsNotOpen" and "enablePsqlAutoSetSearchPath" in the connection configuration class "DBConnectionInfo"

## 3.0.0

- update to Null safety and fix bugs on whereGroup and fix tests

# 3.0.1

- fix bugs on Complex selection With whereGroup, whereSafe, where and whereRaw

# 3.1.1

- fix bug on count query

# 3.2.1

- fix bugs on transaction

# 3.3.1

- add substitutionValues to raw method ``` await db.raw('INSERT INTO products (name,price) VALUES (@name, @price)',  substitutionValues: {'name': 'iPhone 6S', 'price': 2.50}).exec(); ```

# 3.4.1

- add useSSL option to connect when SSL is required

# 3.4.2

- change default timeoutInSeconds to 120

# 3.4.3

- fix bug on query timeout

# 3.5.3

- bug fix when using a combination of whereGroup, whereRaw, orWhereSafe and where

# 4.0.0

- Null security improvements and bug fixes. Attention breack Change: removal of Sanitizer and Validator in all SQL statements and clauses supported in this lib to avoid bugs, now it's up to the lib user to escape invalid fields and table names.

# 5.0.0

- general refactoring, to put double quotes " for Postgresql and ` for MySQL in the table field names for Insert and Update and.
- BREAK CHANGE: now you have to pass the DBConnectionInfo class instance to the DbLayer class constructor, optionally you can override DBConnectionInfo in connect() method.
- bug fix in automated tests

# 5.5.0
- fix bugs
- BREAK CHANGE: the "where" method now puts the quotes envolving the parameter value automatically
    OLD: where('nome ilike ?', "'%isaque%'") NEW: where('nome ilike ?', '%isaque%')
    change optional parameters to named parameters of whereRaw
    whereRaw('b.info ilike @info', andOr: 'AND', substitutionValues: {'info': "%Sant'Ana%"})

example/example.dart
import 'dart:io';
import 'package:fluent_query_builder/fluent_query_builder.dart';

void main() async {
  print('start execution');
  //PostgreSQL connection information
  final pgsqlComInfo = DBConnectionInfo(
    enablePsqlAutoSetSearchPath: true,
    reconnectIfConnectionIsNotOpen: true,
    host: 'localhost',
    database: 'banco_teste',
    driver: ConnectionDriver.pgsql,
    port: 5434,
    username: 'sisadmin',
    password: 's1sadm1n',
    charset: 'utf8',
    schemes: ['jubarte', 'public'],
    setNumberOfProcessorsFromPlatform: false,
    numberOfProcessors: 8,
  );

  var db;
  try {
    print('try connect');
    db = await DbLayer(pgsqlComInfo).connect();
  } catch (e, s) {
    print('catch connect $e $s');
  }

  /*Timer.periodic(Duration(milliseconds: 3000), (timer) async {
    try {
      print('Print after 3 seconds');
       await pgsql
          .select()
          .from('usuarios')
          .whereSafe('login', '=', 'isaque.neves')
          .whereSafe('"idSistema"', '=', '1')
          .getAsMap()
          .then((result) => print('pgsql select $result'));*/

  /*var result = await pgsql.select().from('pessoas').whereGroup((QueryBuilder qb) {
    //return qb.where('nome ilike ?', "'%isaque%'", 'or').where('telefone ilike ?', "'%05%'", 'or');
    return qb.orWhereSafe('nome', 'ilike', '%isaque%').orWhereSafe('telefone', 'ilike', '%05%');
  }).get();*/

  var data = <String, dynamic>{'nome': 'transaction', 'telefone': 'test'};
  var response;
  await db.transaction((ctx) async {
    response = await ctx
        .insertGetAll(returningFields: ['nome', 'telefone'])
        .into('pessoas')
        .setAll(data)
        .exec();
    response = await ctx
        .insertGetAll(returningFields: ['nome', 'telefone'])
        .into('pessoas5')
        .setAll(data)
        .exec();
  });

  print(response);

  //var result = await pgsql.select().from('pessoas').whereRaw("nome ilike '%isaque%'").limit(1).getAsMap();
  // print('pgsql select ${result[0] is Map}');
  exit(0);
  /* 
    } catch (e, s) {
      print('catch select $e s $s');
    }
  });*/

  /* var mysql;
  try {
    print('try connect');
    mysql = await DbLayer().connect(mysqlComInfo);
  } catch (e, s) {
    print('catch connect $e');
  }

  Timer.periodic(Duration(milliseconds: 600), (timer) async {
    try {
      print('Print after 3 seconds');
      await mysql
          .select()
          .from('pessoas')
          .whereSafe('nome', 'like', '%Sant\'Ana%')
          .getAsMap()
          .then((result) => print('mysql select $result'));
    } catch (e, s) {
      print('catch select $e');
    }
  });*/

  /*DbLayer().connect(mysqlCom).then((db) {
    //mysql insert
    /*db
        .insertGetId()
        .into('pessoas')
        .set('nome', 'Isaque Neves Sant\'Ana')
        .set('telefone', '(22) 2771-6265')
        .exec()
        .then((result) => print('mysql insert $result'));*/

    //mysql insertGetId with setAll
    db
        .insertGetId()
        .into('pessoas')
        .setAll({
          'nome': 'Jon Doe',
          'telefone': '171171171',
        })
        .exec()
        .then((result) => print('mysql insertGetId $result'));

    //mysql update with setAll
    db
        .update()
        .whereSafe('id', '=', 13)
        .table('pessoas')
        .setAll({
          'nome': 'Jon Doe',
          'telefone': '171171171',
        })
        .exec()
        .then((result) => print('mysql update $result'));

    //mysql select
    db
        .select()
        //.fields(['login', 'idSistema', 's.sigla'])
        //.fieldRaw('SELECT COUNT(*)')
        .from('pessoas')
        .whereSafe('nome', 'like', '%Sant\'Ana%')
        //.limit(1)
        .getAsMap()
        .then((result) => print('mysql select $result'));
  });*/

  /*DbLayer().connect(mysqlCom).then((db) {
    //mysql insert
    db
        .insert()
        .into('pessoas')
        .set('nome', 'Isaque Neves Sant\'Ana')
        .set('telefone', '(22) 2771-6265')
        .exec()
        .then((result) => print('mysql insert $result'));

    //mysql update
    db
        .update()
        .table('pessoas')
        .set('nome', 'João')
        .where('id=?', 13)
        .exec()
        .then((result) => print('mysql update $result'));

    //mysql delete
    db.delete().from('pessoas')
    .where('id=?', 14)
    .exec()
    .then((result) => print('mysql delete $result'));

    //mysql select
    db
        .select()
        //.fields(['login', 'idSistema', 's.sigla'])
        //.fieldRaw('SELECT COUNT(*)')
        .from('pessoas')
        .whereSafe('nome', 'like', '%Sant\'Ana%')
        //.limit(1)
        .firstAsMap()
        .then((result) => print('mysql select $result'));

    //mysql raw query SELECT * FROM `pessoas` or SELECT COUNT(*) FROM pessoas
    db
        .raw("SELECT * FROM `pessoas`")
        .firstAsMap()
        .then((result) => print('mysql raw $result'));

    //mysql count records
    db
        .select()
        .from('pessoas')
        .orWhereSafe('nome', 'like', '%Sant\'Ana%')
        .orWhereSafe('id', '<', 20)
        .count()
        .then((result) => print('mysql select $result'));
  });*/

  //
  //pgsql insertGetAll
  /* db
      .insertGetAll()
      .into('usuarios')
      .set('username', 'isaque')
      .set('password', '123456')
      .exec()
      .then((result) => print('pgsql insertGetAll $result'));

  db.select().from('usuarios')
  .count()
  .then((result) => print('pgsql count $result'));*/

  /* var data = await pgsql
      .select()
      .from('pessoas')
      // .whereSafe('nome', 'ilike', '%Sant\'Ana%')
      .orWhereGroup((query) {
        return query.orWhereSafe('nome', 'ilike', '%5%').orWhereSafe('cpf', 'ilike', '%5%');
      })
     // .whereSafe('id', '>', 0)
     // .where('id>?', '0')
      .getAsMap();

  print('pgsql select \r\n ${data}');*/

  /*data = await db.getRelationFromMaps(data, 'usuarios', 'idPessoa', 'id');

  

  // var r = await db.select().from('pessoas').fieldRaw('1').limit(1).exec();
  // var r = await db.raw('select 1').exec();
  // print(r);

  /*await db.transaction((ctx) async {

    var result = await ctx.insert().into('usuarios')
    .set('username', 'isaque')
    .set('password', '123456')
    .exec();

    await ctx
        .select()
        .from('pessoas')
        // .whereSafe('nome', 'ilike', '%Sant\'Ana%')
        .orWhereGroup((query) {
          return query
          .orWhereSafe('nome', 'ilike', '%5%')
          .orWhereSafe('cpf', 'ilike', '%5%');
        })
        .whereSafe('id', '>', 0)
        .getAsMap();

    print('pgsql transaction $result');
  });*/

  /*DbLayer().connect(pgsqlCom).then((db) {
    final query = db
        .select()
        //.fields(['login', 'idSistema', 's.sigla'])
        //.fieldRaw('DISTINCT jubarte.sistemas.sigla as')
        //.from('usuarios', alias: 't')
        //  .leftJoin('sistemas', 's.id', '=', 't."idSistema"', alias: 's')
        // .whereRaw("login='isaque.neves'")
        // .whereRaw("s.id='8'")
        // .where("login=?", 'isaque.neves')
        /*.group('login')
      .group('t.idSistema')
      .group('sistemas.sigla');*/
        //.groupRaw('"login", "t"."idSistema", "s"."sigla"')
        .limit(1);
    // .groups(['login', 't.idSistema', 's.sigla']);

    query.firstAsMap().then((onValue) {
      print(onValue);
    });
  });*/
/*
  //example
  var db = await DbLayer(factories: [
    {Usuario: (x) => Usuario.fromMap(x)},
    {Pessoa: (x) => Pessoa.fromMap(x)}
  ]).connect(pgsqlCom);

  //insert Usuario e pessoa relacionada
  var user = Usuario(username: 'jon.doe', password: '123456', idPerfil: 3);
  user.pessoa = Pessoa(nome: 'jon doe', telefone: '717171', cpf: '123');

  //var result = await db.putSingleGetId<Usuario>(user);
  var result = await db.select().from(Usuario.TABLE_NAME).where('id>?', 2).fetchAll<Usuario>();

  print('insert Usuario $result');
  //update Usuario
  //await db.update().where('id=?', 20).updateSingle<Usuario>(Usuario(username: 'jon.doe', password: '987'));
  //select Usuario
 
  //delete Usuario
  //await db.delete().deleteSingle<Usuario>(Usuario(id: 20, username: 'jon.doe', password: '123456'));

  print('end execution');
  // exit(0);*/
}
*/
}
/*
class Usuario implements FluentModelBase {
  Usuario({this.id, this.username, this.password, this.idPerfil, this.idPessoa});

  Usuario.fromMap(Map<String, dynamic> map) {
    id = map['id'];
    username = map['username'];
    password = map['password'];
    ativo = map['ativo'];
    idPerfil = map['idPerfil'];
    idPessoa = map['idPessoa'];
    if (map.containsKey('pessoa')) {
      pessoa = Pessoa.fromMap(map['pessoa']);
    }
  }

  int id;
  String username;
  String password;
  bool ativo;
  int idPerfil;
  int idPessoa;
  Pessoa pessoa;

  @override
  Map<String, dynamic> toMap() {
    final data = <String, dynamic>{};
    if (id != null) {
      data['id'] = id;
    }
    data['username'] = username;
    data['password'] = password;
    data['ativo'] = ativo;
    data['idPerfil'] = idPerfil;
    data['idPessoa'] = idPessoa;
    if (pessoa != null) {
      data['pessoa'] = pessoa.toMap();
    }
    return data;
  }

  static const String TABLE_NAME = 'usuarios';
  static const String ID_KEY = 'id';
  static const String USERNAME_KEY = 'username';
  static const String PASSWORD_KEY = 'password';
  static const String ATIVO_KEY = 'ativo';
  static const String ID_PERFIL_KEY = 'idPerfil';
  static const String ID_PESSOA_KEY = 'idPessoa';

  @override
  OrmDefinitions get ormDefinitions {
    return OrmDefinitions(
      tableName: TABLE_NAME,
      primaryKey: ID_KEY,
      relations: [
        OrmRelation(Pessoa.TABLE_NAME, 'idPessoa', 'id', OrmRelationType.belongsTo, 'pessoa'),
      ],
    );
  }

  @override
  String toString() {
    return toMap().toString();
  }
}

class Pessoa implements FluentModelBase {
  Pessoa({this.id, this.nome, this.telefone, this.cpf});

  Pessoa.fromMap(Map<String, dynamic> map) {
    id = map['id'];
    nome = map['nome'];
    telefone = map['telefone'];
    cpf = map['cpf'];
  }

  int id;
  String nome;
  String telefone;
  String cpf;

  @override
  Map<String, dynamic> toMap() {
    final data = <String, dynamic>{};
    if (id != null) {
      data['id'] = id;
    }
    data['nome'] = nome;
    data['telefone'] = telefone;
    data['cpf'] = cpf;

    return data;
  }

  static const String TABLE_NAME = 'pessoas';
  static const String ID_KEY = 'id';
  static const String NOME_KEY = 'nome';
  static const String TELEFONE_KEY = 'telefone';
  static const String CPF_KEY = 'cpf';

  @override
  OrmDefinitions get ormDefinitions {
    return OrmDefinitions(
      tableName: TABLE_NAME,
      primaryKey: ID_KEY,
    );
  }

  @override
  String toString() {
    return toMap().toString();
  }
}
*/
